lib/Conversion/StandardToHandshake/MaximizeSSA.cpp

```diff
@@ -0,0 +1,237 @@
+      // to visit it
+      if (predBlock != defBlock)
+        if (auto [_, blockNewlyUsing] = blocksUsing.insert(predBlock);
+            blockNewlyUsing)
+          blocksToVisit.push_back(predBlock);
+    }
+  }
+
+  // Replace all uses of the value with the newly added block arguments
+  SmallVector<Operation *> users;
+  for (auto &use : value.getUses()) {
+    auto *owner = use.getOwner();
+    if (owner->getBlock() != defBlock)
+      users.push_back(owner);
+  }
+  for (auto *user : users)
+    user->replaceUsesOfWith(value, blockToArg[user->getBlock()]);
+
+  return success();
+}
+
+LogicalResult circt::maximizeSSA(Operation *op,
+                                 SSAMaximizationStrategy &strategy,
+                                 PatternRewriter &rewriter) {
+  // Apply SSA maximization on each of the operation's results
+  for (auto res : op->getResults())
+    if (strategy.maximizeResult(res))
+      if (failed(maximizeSSA(res, rewriter)))
+        return failure();
+
+  return success();
+}
+
+LogicalResult circt::maximizeSSA(Block *block,
+                                 SSAMaximizationStrategy &strategy,
+                                 PatternRewriter &rewriter) {
+  // Apply SSA maximization on each of the block's arguments
+  for (auto arg : block->getArguments())
+    if (strategy.maximizeArgument(arg))
+      if (failed(maximizeSSA(arg, rewriter)))
+        return failure();
+
+  // Apply SSA maximization on each of the block's operations
+  for (auto &op : block->getOperations())
+    if (strategy.maximizeOp(&op))
+      if (failed(maximizeSSA(&op, strategy, rewriter)))
+        return failure();
+
+  return success();
+}
+
+LogicalResult circt::maximizeSSA(Region &region,
+                                 SSAMaximizationStrategy &strategy,
+                                 PatternRewriter &rewriter) {
+  // Apply SSA maximization on each of the region's block
+  for (auto &block : region.getBlocks())
+    if (strategy.maximizeBlock(&block))
+      if (failed(maximizeSSA(&block, strategy, rewriter)))
+        return failure();
+
+  return success();
+}
+
+namespace {
+
+struct FuncOpMaxSSAConversion : public OpConversionPattern<func::FuncOp> {
+
+  FuncOpMaxSSAConversion(MLIRContext *ctx) : OpConversionPattern(ctx) {}
+
+  LogicalResult
+  matchAndRewrite(func::FuncOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    LogicalResult conversionStatus = success();
+    rewriter.updateRootInPlace(op, [&] {
+      SSAMaximizationStrategy strategy;
+      if (failed(maximizeSSA(op.getRegion(), strategy, rewriter)))
+        conversionStatus = failure();
+    });
+    return conversionStatus;
+  }
+};
+
+struct MaximizeSSAPass : public MaximizeSSABase<MaximizeSSAPass> {
+public:
+  void runOnOperation() override {
+    auto *ctx = &getContext();
+
+    RewritePatternSet patterns{ctx};
+    patterns.add<FuncOpMaxSSAConversion>(ctx);
```

Comment:
Does it even make sense to use a pattern and a rewrite engine here, or could we just directly dispatch to the according functions?